src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -809,6 +813,514 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+  // if ((objects = scopedValueCache()) != null) {
+  bool cache_null_check(Node* c) {
+    if (c->Opcode() == Op_If) {
+      BoolNode* bol = c->in(1)->as_Bool();
+      Node* cmp = bol->in(1);
+      assert(cmp->Opcode() == Op_CmpP, "only reference comparisons in ScopedValue.get()");
+      Node* in1 = cmp->in(1)->uncast();
+      Node* in2 = cmp->in(2)->uncast();
+      if (cache_null_check_with_input(in1, in2, c)) {
+        return true;
+      }
+      if (cache_null_check_with_input(in2, in1, c)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Pattern matches:
+  // if (objects[n] == this) {
+  bool cache_probe(Node* c) {
+    if (c->Opcode() == Op_If) {
+      BoolNode* bol = c->in(1)->as_Bool();
+      Node* cmp = bol->in(1);
+      assert(cmp->Opcode() == Op_CmpP, "only reference comparisons in ScopedValue.get()");
+      Node* in1 = cmp->in(1)->uncast();
+      Node* in2 = cmp->in(2)->uncast();
+      Node* uncasted_scoped_value_object = _scoped_value_object->uncast();
+      assert(in1 == uncasted_scoped_value_object || in2 == uncasted_scoped_value_object,
+             "one of the comparison inputs must be the scoped value oop");
+      Node* in = in1 == uncasted_scoped_value_object ? in2 : in1;
+      BarrierSetC2* bs = BarrierSet::barrier_set()->barrier_set_c2();
+      in = bs->step_over_gc_barrier(in);
+      if (in->Opcode() == Op_DecodeN) {
+        in = in->in(1);
+      }
+      assert(in->Opcode() == Op_LoadP || in->Opcode() == Op_LoadN, "load from cache array expected");
+      assert(_kit->C->get_alias_index(in->adr_type()) == _kit->C->get_alias_index(TypeAryPtr::OOPS),
+             "load from cache array expected");
+      AddPNode* addp1 = in->in(MemNode::Address)->as_AddP();
+      ProjNode* sv_cache_proj = addp1->in(AddPNode::Base)->uncast()->as_Proj();
+      assert(sv_cache_proj->in(0)->as_CallJava()->method()->intrinsic_id() == vmIntrinsics::_scopedValueCache,
+             "should be call to Thread.scopedValueCache()");
+      assert(_scoped_value_cache == nullptr || _scoped_value_cache == sv_cache_proj->in(0),
+             "only one cache expected");
+      _scoped_value_cache = sv_cache_proj->in(0)->as_Call();
+      assert(in->in(MemNode::Memory)->is_Proj() && in->in(MemNode::Memory)->in(0) == _scoped_value_cache,
+             "load from cache expected right after Thread.scopedValueCache() call");
+      Node* addp2 = addp1->in(AddPNode::Address);
+      Node* offset1 = addp1->in(AddPNode::Offset);
+      intptr_t const_offset = offset1->find_intptr_t_con(-1);
+      BasicType bt = TypeAryPtr::OOPS->array_element_basic_type();
+      int shift = exact_log2(type2aelembytes(bt));
+      int header = arrayOopDesc::base_offset_in_bytes(bt);
+      assert(const_offset >= header, "load from cache doesn't access the cache array?");
+      const_offset -= header;
+
+      Node* index_in_cache_array = _kit->gvn().intcon(const_offset >> shift);
+      if (addp2->is_AddP()) {
```

Comment:
You either need a comment here, or maybe better variable names can help.
Currently, I can't really review this / convince myself you are covering all cases without reverse-engineering all variable names and the code.